File: twoSum/twoSum.js
// In N2 Time complexity

const twoSumFunction = (nums, target) => {
  for (let p = 0; p < nums.length; p++) {
    const numberToFind = target - nums[p];
    for (let j = p + 1; j < nums.length; j++) {
      if (nums[j] === numberToFind) {
        return [p, j];
      }
    }
  }

  return null;
};

const nums = [2, 3, 9, 1, 6];
console.log(twoSumFunction(nums, 11));

// In N Time complexity And N space Complexity
const twoSumUsingHashMapFunction = (nums, target) => {
  const newMap = {};

  for (let p = 0; p < nums.length; p++) {
    const valueInMap = newMap[nums[p]];

    if (valueInMap >= 0) {
      return [valueInMap, p];
    } else {
      const numberTopFind = target - nums[p];
      newMap[numberTopFind] = p;
    }
  }

  return null;
};

console.log(twoSumUsingHashMapFunction(nums, 11));

File: Others/twoSum/twoSum.py
# In N2 Time complexity

def two_sum_function(nums, target):
    for p in range(len(nums)):
        number_to_find = target - nums[p]
        for j in range(p + 1, len(nums)):
            if nums[j] == number_to_find:
                return [p, j]

    return None

def two_sum_using_hashmap_function(nums, target):
    new_map = {}

    for p in range(len(nums)):
        value_in_map = new_map.get(nums[p], -1)

        if value_in_map >= 0:
            return [value_in_map, p]
        else:
            number_to_find = target - nums[p]
            new_map[number_to_find] = p

    return None


print(two_sum_function([2, 3, 9, 1, 6], 11));
print(two_sum_using_hashmap_function([2, 3, 9, 1, 6], 11));
